x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUserSerializationHelper.java

```diff
@@ -36,6 +38,9 @@ public static void writeTo(User user, StreamOutput output) throws IOException {
         } else if (XPackSecurityUser.is(user)) {
             output.writeBoolean(true);
             output.writeString(XPackSecurityUser.NAME);
+        } else if (AsyncSearchUser.is(user)) {
```

Comment:
Is this a concern for bwc? I think, in a mixed cluster environment, writing and reading this information from different versions would fail. But current failures don't seem to be releavant. I don't know the details about async search work. Is it 8.0 only?

Reply:
It's a 7.7 feature so nodes in this version will be able to run `_async_search` and to create the new `.async-search` index. For the few cases where an older node could be  picked, the `IllegalStateException` thrown by `readFrom` should be enough to handle failures gracefully ?

Reply:
This is indeed a problem, but I don't see an easy way around it.

With the proposed changes, async search won't work until the master node is also upgraded, even if the coordinating node is. Does this sound reasonable to you @jimczi ? Are there other issues that would prevent the async search feature to work in a mixed cluster (assuming no, but maybe there is :) )?

Reply:
That's fine IMO. We expect users to upgrade clusters entirely before using new features . Also, Kibana will use `_async_search` internally so the ES cluster must be fully upgraded before upgrading Kibana.